File: src/main/java/edu/marmara/model/Course.java
package edu.marmara.model;

import java.util.List;


public class Course {
    private String courseCode;
    private String courseTitle;
    private Integer givenSemester;
    private List<Student> enrolledStudents;
    private Integer courseCredit;
    private List<Course> prerequisites;
    private Instructor instructor;
    private List<WeeklyDate> dates;
    private Integer maxSeats;
    private Integer takenSeats;
    private Integer minCreditReq;


    public Course() {
    }

    public Course(String courseCode, String courseTitle, Integer givenSemester, List<Student> enrolledStudents, Integer courseCredit, List<Course> prerequisites, Instructor instructor, List<WeeklyDate> dates, Integer maxSeats, Integer minCreditReq) {
        this.courseCode = courseCode;
        this.courseTitle = courseTitle;
        this.givenSemester = givenSemester;
        this.enrolledStudents = enrolledStudents;
        this.courseCredit = courseCredit;
        this.prerequisites = prerequisites;
        this.instructor = instructor;
        this.dates = dates;
        this.maxSeats = maxSeats;
        this.takenSeats = maxSeats;
        this.minCreditReq = minCreditReq;
    }

    public void addWeeklyDate(WeeklyDate weeklyDate){
        this.dates.add(weeklyDate);
    }
    public void addPrerequisite(Course course){
        this.prerequisites.add(course);
    }


    @Override
    public String toString() {
        return "Course{" +
                "courseCode='" + courseCode + '\'' +
                ", courseTitle='" + courseTitle + '\'' +
                ", givenSemester=" + givenSemester +
                ", enrolledStudents=" + enrolledStudents +
                ", courseCredit=" + courseCredit +
                ", prerequisites=" + prerequisites +
                ", instructor=" + instructor +
                ", dates=" + dates +
                '}';
    }

    public void increaseTakenSeat(){ this.takenSeats++; }
    public void decreaseTakenSeat(){ this.takenSeats--; }

    public String getCourseCode() {
        return courseCode;
    }

    public void setCourseCode(String courseCode) {
        this.courseCode = courseCode;
    }

    public String getCourseTitle() {
        return courseTitle;
    }

    public void setCourseTitle(String courseTitle) {
        this.courseTitle = courseTitle;
    }

    public Integer getGivenSemester() {
        return givenSemester;
    }

    public void setGivenSemester(Integer givenSemester) {
        this.givenSemester = givenSemester;
    }

    public List<Student> getEnrolledStudents() {
        return enrolledStudents;
    }

    public void setEnrolledStudents(List<Student> enrolledStudents) {
        this.enrolledStudents = enrolledStudents;
    }

    public Integer getCourseCredit() {
        return courseCredit;
    }

    public void setCourseCredit(Integer courseCredit) {
        this.courseCredit = courseCredit;
    }

    public List<Course> getPrerequisites() {
        return prerequisites;
    }

    public void setPrerequisites(List<Course> prerequisites) {
        this.prerequisites = prerequisites;
    }

    public Instructor getInstructor() {
        return instructor;
    }

    public void setInstructor(Instructor instructor) {
        this.instructor = instructor;
    }

    public List<WeeklyDate> getDates() {
        return dates;
    }

    public void setDates(List<WeeklyDate> dates) {
        this.dates = dates;
    }

    public Integer getTakenSeats() {
        return takenSeats;
    }

    public void setTakenSeats(Integer availableSeats) {
        this.takenSeats = availableSeats;
    }

    public Integer getMaxSeats() {
        return maxSeats;
    }

    public void setMaxSeats(Integer maxSeats) {
        this.maxSeats = maxSeats;
    }

    public Integer getMinCreditReq() {
        return minCreditReq;
    }

    public void setMinCreditReq(Integer minCreditReq) {
        this.minCreditReq = minCreditReq;
    }
}

File: src/python/src/main/marmara/model/Course.py
from typing import List, Union


class Course:
    def __init__(self, course_code: str, course_title: str, given_semester: int, enrolled_students: List["Student"],
                 course_credit: int, prerequisites: List["Course"], instructor: "Instructor", dates: List["WeeklyDate"],
                 max_seats: int, min_credit_req: int):
        self.course_code = course_code
        self.course_title = course_title
        self.given_semester = given_semester
        self.enrolled_students = enrolled_students
        self.course_credit = course_credit
        self.prerequisites = prerequisites
        self.instructor = instructor
        self.dates = dates
        self.max_seats = max_seats
        self.taken_seats = max_seats
        self.min_credit_req = min_credit_req

    def add_weekly_date(self, weekly_date: "WeeklyDate"):
        self.dates.append(weekly_date)

    def add_prerequisite(self, course: "Course"):
        self.prerequisites.append(course)

    def increase_taken_seat(self):
        self.taken_seats += 1

    def decrease_taken_seat(self):
        self.taken_seats -= 1

    def __str__(self):
        return f"Course{{course_code='{self.course_code}', course_title='{self.course_title}', given_semester={self.given_semester}, enrolled_students={self.enrolled_students}, " \
               f"course_credit={self.course_credit}, prerequisites={self.prerequisites}, instructor={self.instructor}, dates={self.dates} "

    @property
    def course_code(self) -> str:
        return self._course_code

    @course_code.setter
    def course_code(self, course_code: str):
        self._course_code = course_code

    @property
    def course_title(self) -> str:
        return self._course_title

    @course_title.setter
    def course_title(self, course_title: str):
        self._course_title = course_title

    @property
    def given_semester(self) -> int:
        return self._given_semester

    @given_semester.setter
    def given_semester(self, given_semester: int):
        self._given_semester = given_semester

    @property
    def enrolled_students(self) -> List["Student"]:
        return self._enrolled_students

    @enrolled_students.setter
    def enrolled_students(self, enrolled_students: List["Student"]):
        self._enrolled_students = enrolled_students

    @property
    def course_credit(self) -> int:
        return self._course_credit

    @course_credit.setter
    def course_credit(self, course_credit:int):
        self._course_credit = course_credit
